### ISSUE TITLE
Make the completion trigger character conditional

### ISSUE THREAD

**Issue by jdneo** [microsoft/language-server-protocol#1413]

**Comment by dbaeumer** [comment]
@jrieken has this come up in VS Code ?

**Comment by hrsh7th** [comment]
The [lua-language-server](https://github.com/sumneko/lua-language-server) achives the similar functionality.
That LS specified the ` ` as triggerCharacter and `textDocument/completion` returns context-aware completion response.

**Comment by jdneo** [comment]
Share one more information. I observed the `ms-dotnettools.csharp` also implemented this feature.

Currently in Java, we are planning to add `" "` as a trigger character and looking backward to see if it's `"new "` at where the completion is triggered.

**Comment by puremourning** [comment]
In ycmd triggers were always regex based. LSP using a single character has always seemed limiting (think, even c `->` and c++ `::`). Adding space as a trigger character is a bad idea imo as semantic completion requests can be quite slow, particularly for ‘global scope’ in many languages. And this is going to lead to a _lot_ of pointless requests. 

I think a protocol supported regex trigger would be better. Or just press a manual trigger.

**Comment by jrieken** [comment]
> @jrieken has this come up in VS Code ?

Yes, it has come up in the past but not in recent years. I believe extension use the last character of their words like `>` of `->` and check the prefix themselves. We can consider adding support for longer prefix strings but I am unsure about regex because it comes at a certain cost.

**Comment by hrsh7th** [comment]
I'm against regular expressions. I think there are many dialects and some clients cannot handle it.

The advantage of a single character is that it can be handled by the editor's keypress handler.
However, if we make the space a trigger string, we may consider that the completion request runs indented...

**Comment by puremourning** [comment]
FWIW, these are the regex triggers that ycmd uses by default. We add the 'trigger characters' provided by LSP servers to these lists dynamically: https://github.com/ycm-core/ycmd/blob/master/ycmd/completers/completer_utils.py#L180-L211 (for some languages we actually ignore them, as they are inferior in practice to our hard-coded ones).

The LSP already includes a regular expression syntax, defined to be some subset of javascript's (https://microsoft.github.io/language-server-protocol/specifications/specification-current/#regExp). Anyway, I do tend to agree that is somewhat unfortunate and in practice (as you see above) most expression are just "multiple characters" rather than actually complex expressions.

Perhaps the 'trigger characters' could be augmented with alternative "trigger prefixes" which is to say "Trigger if the prefix at the current cursor position matches these characters". The challenge there is that the "whitespace" trigger that prompted this request no longer works if you enter multiple spaces as in `X x = new               <here>` (whereas it's trivial with a regex)
